internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,165 @@
+	// Holds the incoming data from kernel write calls, currently being buffered
+	// for upload in the future.
+	current inMemoryBuffer
+
+	// Holds the data that is currently being uploaded asynchronously to GCS. Data
+	// in this buffer will persist until the entire data is successfully uploaded.
+	flushed inMemoryBuffer
+
+	// Holds the last offset to which data is written.
+	fileSize int64
+
+	// Holds the buffer's size (in bytes) to be allocated to currentBuffer and flushedBuffer.
+	bufferSize int64
 }
 
-// CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
-// set to nil. Memory is allocated to the buffer when the first write call comes.
-// This avoids unnecessarily bloating GCSFuse memory consumption.
-//
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
-func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
-	b := &InMemoryWriteBuffer{}
+// CreateInMemoryWriteBuffer returns a new InMemoryWriteBuffer.
+// Memory is lazily allocated to current and flushed buffer on their first usage.
+func CreateInMemoryWriteBuffer(bufferSizeMB uint) (*InMemoryWriteBuffer, error) {
+	if bufferSizeMB == 0 {
+		return nil, fmt.Errorf(ZeroSizeBufferError)
+	}
+	b := &InMemoryWriteBuffer{
+		bufferSize: int64(bufferSizeMB * MiB),
+	}
 	// TODO: set mtime attribute.
-	return b
+	logger.Debugf("TODO: set mtime attribute.")
+	return b, nil
+}
+
+// Allocates memory to currentBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureCurrentBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.current.buffer == nil {
+		b.current.buffer = make([]byte, 0, b.bufferSize)
+		b.current.offset.end = b.bufferSize
+	}
+	return nil
+}
+
+// Allocates memory to flushedBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureFlushedBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.flushed.buffer == nil {
+		b.flushed.buffer = make([]byte, 0, b.bufferSize)
+	}
+	return nil
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
+	contentSize := int64(len(content))
+	if contentSize == 0 {
+		return nil
+	}
+
+	// Ensure b.currentBuffer != nil
+	if err := b.ensureCurrentBuffer(); err != nil {
+		return fmt.Errorf("ensureCurrentBuffer: %w", err)
+	}
+
+	if offset < b.current.offset.start ||
+		offset >= b.current.offset.end+b.bufferSize {
+		// TODO: finalise write and trigger temp file flow.
+		logger.Debugf("Non-sequential write encountered. TODO: Switch to " +
+			"temp-file flow insteading of erroring out.")
+		return fmt.Errorf(NonSequentialWriteError)
 	}
+
+	var contentWrittenSoFar int64
+	for {
+		n := b.writePartialContentToBuffer(content[contentWrittenSoFar:], offset+contentWrittenSoFar)
```

Comment:
@ashmeenkaur  I see a bit of a tricky situation here, which might cause a deadlock.

The updated checks on the limits on the value of `offset` allow the passed `offset` to be in either the current `current` buffer or in the next `current` buffer.

Assuming that the current `current` buffer is not full yet, and the passed `offset` is in `[current.offset.end, current.offset+bufferSize)` range, then on the first iteration of this for-loop, the returned `n` will be 0 and each iteration will just do the same.

Now, contentWrittenSoFar will never be increased, and the terminating condition `contentWrittenSoFar == int64(len(content))` will never be attained.

I might be missing something. Please check and let me know.